src/Microsoft.Diagnostics.Runtime/src/Windows/AWEBasedCacheEntry.cs

```diff
@@ -132,9 +132,22 @@ public override long PageOutData()
             return sizeRemoved;
         }
 
-        protected override void InvokeCallbackWithDataPtr(CachePage<UIntPtr> page, Action<UIntPtr, uint> callback)
+        protected unsafe override uint InvokeCallbackWithDataPtr(CachePage<UIntPtr> page, Func<UIntPtr, uint, uint> callback)
         {
-            callback(page.Data, page.DataExtent);
+            return callback(page.Data, page.DataExtent);
+        }
+
+        protected override uint CopyDataFromPage(CachePage<UIntPtr> page, IntPtr buffer, uint inPageOffset, uint byteCount)
+        {
+            // Calculate how much of the requested read can be satisfied by the page
+            uint sizeRead = Math.Min(page.DataExtent - inPageOffset, byteCount);
+
+            unsafe
+            {
+                CacheNativeMethods.Memory.memcpy(buffer, new UIntPtr((byte*)page.Data + inPageOffset), new UIntPtr(sizeRead));
```

Comment:
This doesn't have to be changed for this checkin though.

Reply:
Sorry, just saw these questions.

Previously the call to read data (the standard route), looked like this:

ReadPageDataFromOffset(pageIndex, (data, dataLength) =>
                {
                    // Calculate how much of the requested read can be satisfied by the page
                    sizeRead = Math.Min(dataLength - inPageOffset, byteCount);

                    unsafe
                    {
                        CacheNativeMethods.Memory.memcpy(buffer, new UIntPtr((byte*)data + inPageOffset), new UIntPtr(sizeRead));
                    }
                });

So ReadPageDataFromOffset took an Action, I passed in a lambda here.

The creation of the lambda object + the creation of the Action wrapper around the lambda were very short lived, since once this method returned they were GC eligible.

In my perf analysis of my platform using this code around 23% of GCed objects in a run were that lambda + the Action. That seemed a bit high for such a hot path.

The new code basically encapsulates what was in the lambda into a virtual function call that is made inside ReadPageDataFromOffset, each cache entry implements it to copy the data, because each cache entry has a separate data backing store (IntPtr and byte[]). 

So now ReadPageDataFromOffset still takes a callback, but it can be null (and is in the common case), if it is null it assumes it is a 'simple copy' and invokes the virtual function to do that. For more complex copy semantics (like CopyUntil where it has to look for a terminating sequence and possibly 'carry over' read data between pages) it still uses the user provided callback, but that is like the 0.001% case. 

Thus there should be no temporary lambdas or Action objects, lessening the GC pressure. I noticed a noticeable wall-clock improvement in my processing of dumps that had > 50 million objects on the managed heap (the one that had 75 million, this one change shaved about 40 seconds off the runtime end to end).

Reply:
> These should really be using `Span<T>` throughout and `Span.CopyTo`. I know it's a bit of a newer feature to learn but it makes a big difference in debugability and safety. Perf is on par with using native pointers, but the GC can really track the pointer throughout and it makes things a lot better in the library.

Yes, ultimately I want to sweep through my codebase making that change as it uses IntPtr tons of place for serialization perf, I would love to move them all to Span<T>, where I can, some may need to stay IntPtr as they deal with buffer passed to and returned from Win32 API calls, which have no clue of Span<T> obviously.